Tests/AsyncQueueTests/SemaphoreTests.swift

```diff
@@ -0,0 +1,144 @@
+                    systemUnderTest.signal()
+                }
+            }
+        }
+
+        // Loop one fewer than iterationCount.
+        for _ in 1..<iterationCount {
+            await systemUnderTest.execute { systemUnderTest in
+                systemUnderTest.signal()
+            }
+        }
+
+        await systemUnderTest.execute { systemUnderTest in
+            // Give each suspended `wait` task an opportunity to resume (if they were to resume, which they won't) before we check the count.
+            for _ in 1...iterationCount {
+                await Task.yield()
+            }
+
+            // The count will still be zero each time because we have executed one more `wait` than `signal` calls.
+            let completedCountedTasks = unsafeCounter.countedTasksCompleted
+            XCTAssertEqual(completedCountedTasks, 0)
+
+            // Signal one last time, enabling all of the original `wait` calls to resume.
+            systemUnderTest.signal()
+
+            for _ in 1...iterationCount {
+                // Wait for all enqueued `wait`s to have completed and signaled their completion.
+                await systemUnderTest.wait()
+            }
+
+            let tasksCompleted = unsafeCounter.countedTasksCompleted
+            XCTAssertEqual(iterationCount, tasksCompleted)
+        }
+    }
+
+    func test_wait_doesNotSuspendIfSignalCalledFirst() async {
+        await systemUnderTest.signal()
+        let didSuspend = await systemUnderTest.wait()
+        XCTAssertFalse(didSuspend)
+    }
+
+    // MARK: Private
+
+    private var systemUnderTest = Semaphore()
+}
+
+// MARK: - Semaphore Extension
+
+private extension Semaphore {
+    /// Enqueues an asynchronous task. This method suspends the caller until the asynchronous task has begun, ensuring ordered execution of enqueued tasks.
+    /// - Parameter task: A unit of work that returns work to execute after the task completes and the count is incremented.
+    func enqueueAndCount(using counter: UnsafeCounter, _ task: @escaping @Sendable (isolated Semaphore) async -> ((isolated Semaphore) -> Void)?) async {
+        // Await the start of the soon-to-be-enqueued `Task` with a continuation.
+        await withCheckedContinuation { continuation in
+            // Re-enter the actor context but don't wait for the result.
+            Task {
+                // Now that we're back in the actor context, resume the calling code.
+                continuation.resume()
+                let executeAfterIncrement = await task(self)
+                counter.countedTasksCompleted += 1
+                executeAfterIncrement?(self)
+            }
+        }
+    }
+
+    func execute(_ task: @Sendable (isolated Semaphore) async throws -> Void) async rethrows {
+        try await task(self)
+    }
+}
+
+// MARK: - UnsafeCounter
+
+private final class UnsafeCounter: @unchecked Sendable {
+    var countedTasksCompleted = 0
+}
```

Comment:
It's OK that this isn't thread-safe since it's only accessed on the actor's context. Is that right?

Reply:
Correct!